[24.0 backport] image/cache: Ignore Build and Revision on Windows

vvoland opened the pull request:
- backport: https://github.com/moby/moby/pull/47330
- backport: https://github.com/moby/moby/pull/47342
- fixes: https://github.com/moby/moby/issues/47307

The compatibility depends on whether `hyperv` or `process` container isolation is used.
This fixes cache not being used when building images based on older Windows versions on a newer Windows host.

**- How to verify it**

**- Description for the changelog**
```release-note
- Windows: Fix cache not being used when building images based on Windows versions older than the host's version
```

**- A picture of a cute animal (not mandatory but encouraged)**

vvoland committed:
image/cache: Ignore Build and Revision on Windows

The compatibility depends on whether `hyperv` or `process` container
isolation is used.
This fixes cache not being used when building images based on older
Windows versions on a newer Windows host.

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>
(cherry picked from commit 91ea04089b75f3c56e60e33e27daeb0d494d79f5)
Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

thaJeztah reviewed (approved):
LGTM

thaJeztah commented:
I think we're missing an import in this branch;

```
=== Errors
image/cache/compare_test.go:135:22: undefined: platforms.Platform
image/cache/compare_test.go:136:22: undefined: platforms.Platform
image/cache/compare_test.go:141:24: undefined: platforms.Platform
image/cache/compare_test.go:142:24: undefined: platforms.Platform
image/cache/compare_test.go:147:24: undefined: platforms.Platform
image/cache/compare_test.go:148:24: undefined: platforms.Platform
image/cache/compare_test.go:153:24: undefined: platforms.Platform
image/cache/compare_test.go:154:24: undefined: platforms.Platform
image/cache/compare_test.go:159:24: undefined: platforms.Platform
image/cache/compare_test.go:160:24: undefined: platforms.Platform
image/cache/compare_test.go:160:24: too many errors
```

vvoland committed:
image/cache: Use Platform from ocispec

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>
(cherry picked from commit 2c01d53d9692250a7d1b18742f1e100e43615586)
Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

vvoland commented:
Dang, looks like I accidentally used a `Platform` type alias from containerd instead of the `ocispec.Platform` struct directly, and it's not available on older containerd.

I opened https://github.com/moby/moby/pull/47342 to use ocispec instead on master to make the code consistent and will also include it in these backport PRs.

vvoland commented:
Weird, looks like some unit test cases fail on the 24 and 23:

```
=== Failed
=== FAIL: github.com/docker/docker/image/cache TestPlatformCompare/different_major (0.01s)
    compare_test.go:203: assertion failed: true (bool) != false (tc.expected bool)

=== FAIL: github.com/docker/docker/image/cache TestPlatformCompare/different_minor_same_osver (0.00s)
    compare_test.go:203: assertion failed: true (bool) != false (tc.expected bool)

```

thaJeztah commented:
Could it be different version of oci-spec or different compare in containerd? (haven''t checked the test yet)

vvoland commented:
Looks like the containerd platform matcher in 1.6 doesn't have the proper OSVersion check: https://github.com/moby/moby/blob/24.0/vendor/github.com/containerd/containerd/platforms/defaults_windows.go

vvoland committed:
image/cache: Require Major and Minor match for Windows OSVersion

The platform comparison was backported from the branch that vendors
containerd 1.7.

In this branch the vendored containerd version is older and doesn't have
the same comparison logic for Windows specific OSVersion.

Require both major and minor components of Windows OSVersion to match.

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

thaJeztah reviewed (approved):
LGTM
